src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -281,16 +289,24 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
                 if self.mir.local_kind(index) == LocalKind::Temp
                 && self.temp_promotion_state[index].is_promotable() {
                     debug!("store to promotable temp {:?}", index);
-                    store(&mut self.temp_qualif[index]);
+                    store(&mut self.local_qualif[index]);
                 }
             }
             return;
         }
 
         match *dest {
+            Place::Local(index) if self.mir.local_kind(index) == LocalKind::Var &&
+                                   self.tcx.sess.features_untracked().const_let => {
+                debug!("store to var {:?}", index);
+                match self.local_qualif[index] {
+                    Some(ref mut qualif) => *qualif &= self.qualif,
```

Comment:
I think you want to replace the `qualif` always (e.g. `let mut x = None; x = Some(TypeThatNeedsDrop);`).
Also, if you have this special case, why do you need to do anything about `statement_like`?